src/interceptor.ts

```diff
@@ -44,6 +45,11 @@ export const aws4Interceptor = (options?: InterceptorOptions, credentials?: Cred
     throw new Error("No URL present in request config, unable to sign request");
   }
 
+  if(config.params) {
+    config.url = buildUrl(config.url, config.params, config.paramsSerializer);
```

Comment:
Not sure if `buildUrl` does the alphabetical sort, but it is part of the signing procedure.

https://docs.aws.amazon.com/general/latest/gr/sigv4-create-canonical-request.html

> Sort the parameter names by character code point in ascending order. Parameters with duplicate names should be sorted by value. For example, a parameter name that begins with the uppercase letter F precedes a parameter name that begins with a lowercase letter b.

Reply:
buildUrl does not sort. I see it's stated in the docs. Testing against the Amazon seller api, which requires AWS4 signature, query param order made no difference. Signature matched, regardless param order.

Reply:
It definitely makes a difference. I have ran into it, but I don't remember with which call.

I have sorting added to my hack to make this work.

Reply:
It does not make a difference, since aws4 already sorts query params according to the spec. https://github.com/mhart/aws4/blob/a413aadd9e4b4e58842937a9ad53354be41ef4a1/aws4.js#L260

Reply:
The parsing of the object happens here:

https://github.com/mhart/aws4/blob/a413aadd9e4b4e58842937a9ad53354be41ef4a1/aws4.js#L123-L125

And it depends on the following param being set: `signQuery`.

But this package does not set this param, and thus `query` does not get parsed and does not get sorted.

Reply:
No, I don't think this is correct. `RequestSigner.prototype.canonicalString` is called regardless whether signQuery is set to true or not, as is the sorting of the query params. Could you please check the execution flow? 

`Sign` calls `authHeader()` calls `signature()` calls `stringToSign()` calls `canonicalString()` which does the sorting.

Reply:
I made a request and observed the traffic flow using Charles (MITM proxy).

Request:

```ts
  const orders = await request({
    method: 'GET',
    url: '/orders/v0/orders',
    params: {
      MarketplaceIds: MARKETPLACE_ID,
      CreatedAfter: '2020-01-01T00:00:00',
    },
  })
```

Charles HTTP log:

```http
GET /orders/v0/orders?MarketplaceIds=A1805IZSGTT6HS&CreatedAfter=2020-01-01T00%3A00%3A00 HTTP/1.1
```

As you can see the params did not get sorted.

The request still succeeded, but it is not guaranteed it always would. Also, as I can tell, different endpoints were designed by different teams (naming conventions are different), so it is quite possible that some endpoints will enforce sorting and others wouldn't.

In any case, if the signing standard calls for sorting, I think we should be sorting, even if it is not yet required.

---

![image](https://user-images.githubusercontent.com/491247/103721395-1fe11980-5000-11eb-9f5a-f9d9018a83d8.png)

Reply:
Alright, I was under the impression that the sorting only needs to happen for creating the signature. Shall we borrow implementation from aws4 then? https://github.com/mhart/aws4/blob/a413aadd9e4b4e58842937a9ad53354be41ef4a1/aws4.js#L252-L269

Reply:
> I was under the impression that the sorting only needs to happen for creating the signature.

Oh I see. So you think that the `aws4` code will sort and sign it. And then when Amazon receives the request, they won't try to sign the URI as-is, but they'll parse it, and sort and sign also?

In other words what goes into the actual HTTP request does not matter?

Reply:
That was my assumption. But it is solely based on good faith in the aws4 author. If the actual http request would matter, implementing query param sorting in a signing lib like aws4 would make no sense to me. It would lead to missmatch errors all the time – like in our case where the url for signing would be correctly sorted but the actual url is not.

But truth be told: After googling about the signing processes it seems that different aws teams indeed implement the signing in various ways and I found a few issues that pointed out that the aws sdk tests are not 100% compliant with the spec as well. So I think you are right, it should be implemented.

Reply:
Also see #35. Not 100% related to this issue, but somewhat is.

Reply:
Okay, probably better to close this PR in favor of #35

Reply:
I think #35 is complimentary to this one. As `@aws-sdk/signature-v4` replaces `aws4` package. But we still need to handle the params.